feature(ACE): Handle variable length public inputs and generate powers of randomness

adr1anh opened the pull request:
Fixes #399 and #397 

**Changes**:
- Update the ACE codegen to handle bus boundary constraints which refer to public input tables.
- Update `ir::Air` to return the list of public table accesses with a canonical order.
  - Use this accessor to simplify winterfell codegen
- Improve layout of ACE inputs to satisfy requirements of the MASM verifier.
- Fix random challenges to α and β and derive powers of the latter in the circuit.

adr1anh committed:
Fixed #414

*Introduce new `ir::PublicInputTable` node*: This node corresponds to a randomly-reduced public input table.

*Optimize Winterfell bus boundary constraint generation*: Using the new `PublicInputTable` node, we can compute the reduced public input tables once, and handle all bus boundary constraints uniformly

adr1anh committed:
CHANGELOG.md

adr1anh committed:
Remove `bus_name` which could lead to duplicate public input evaluations.

adr1anh committed:
Initial commit

adr1anh committed:
- Make random challenges α, β explicit
- More granular input layout
- Document

Al-Kindi-0 commented on codegen/ace/src/builder.rs:
```diff
@@ -306,6 +308,25 @@ impl CircuitBuilder {
     }
+
+    /// Returns a [`Node`] corresponding to the random challenge at the given index.
+    /// We assume that the challenges are [ β, 1, α, α², α³, … ].
```
Would be great if we can flip alpha and beta i.e., `[alpha, 1, beta, beta^2, ...]`. I think this ordering was suggested by the MASM code but the reason alpha and beta are out of order there is mainly for optimization purposes

adr1anh replied:
Done

Al-Kindi-0 commented on codegen/ace/src/builder.rs:
```diff
@@ -306,6 +308,25 @@ impl CircuitBuilder {
+        if index == 0 {
+            return self.layout.random_beta_node();
+        }
+        // TODO: Does rand[1] equal 1, or alpha? We assume 1 here.
```
Indeed, it should be one

Al-Kindi-0 commented on codegen/ace/src/inputs.rs:
```diff
@@ -18,10 +18,14 @@ pub struct AirInputs {
     pub main: [Vec<QuadFelt>; 2],
-    /// Verifier challenges used to derive the *aux* trace.
-    pub rand: Vec<QuadFelt>,
+    /// Verifier challenge α used to fingerprint bus messages for the *aux* trace.
```
nit: Assuming the order alpha then beta, then alpha is used for the multi-set or logup argument i.e., $X \leftarrow \alpha$. $\beta$ is then used to fingerprint/reduce the messages.

Al-Kindi-0 commented on codegen/ace/src/layout.rs:
```diff
@@ -56,21 +66,51 @@ impl Layout {
+            *offset += offset.next_multiple_of(alignment);
+        }
+
+        // TODO: should these be element or word aligned?
```
(fixed length) public inputs should be double word aligned as a whole. This means that within (fixed length) PI we should have element-alignment
variable length public inputs should also be double word aligned per message. This would make Horner evaluation more optimized.

Al-Kindi-0 commented on codegen/ace/src/layout.rs:
```diff
@@ -56,21 +66,51 @@ impl Layout {
     /// Returns a new [`Layout`] from a description of an [`Air`]. All regions are padded according
     /// to `HASH_ALIGNMENT`, ensuring that each section starts at a word-aligned memory pointer.
```
should probably be updated

Al-Kindi-0 commented on codegen/ace/src/layout.rs:
```diff
@@ -56,21 +66,51 @@ impl Layout {
+
+        // Region containing all reduced public input table values.
+        let reduced_tables_region =
+            next_region(offset, reduced_table_accesses.len(), WORD_ALIGNMENT);
```
Could we store one reduced table per word? This simplifies the MASM codegen work significantly

Al-Kindi-0 commented on codegen/ace/src/layout.rs:
```diff
@@ -98,17 +138,29 @@ impl Layout {
+        let stark_vars = next_region(offset, StarkVar::num_vars(), WORD_ALIGNMENT);
+
+        // Ensure the entire input region is double-word aligned
+        // TODO: Not sure if this is actually needed by the MASM verifier.
```
don't think so either

Al-Kindi-0 reviewed (commented):
Did an initial review and it LGTM!

adr1anh committed:
Address feedback

- Swap roles of alpha and beta
- Update alignment rules to match MASM verifier.
- Document difference between alpha (multiset/logUp challenge) and beta (fingerprinting)

Al-Kindi-0 commented on air/src/ir/mod.rs:
```diff
@@ -103,6 +103,29 @@ impl Air {
         self.public_inputs.values()
     }
 
+    /// Returns a list of all access to reduced public input tables in canonical order.
```
nit: access -> accesses ?

Al-Kindi-0 commented on codegen/ace/src/layout.rs:
```diff
@@ -40,37 +53,71 @@ pub struct Layout {
             .collect();
 
-        let random_values = next_region(&mut inputs_offset, air.num_random_values as usize);
+        // Random challenges α, β used to fingerprint bus messages.
```
nit: // Random challenges α, β used to fingerprint bus messages. -> // Random challenges α, β.

adr1anh committed:
Merge branch 'next' into ace-var-pi

# Conflicts:
#	air-script/tests/buses/buses_varlen_boundary_both.rs
#	air-script/tests/buses/buses_varlen_boundary_first.rs
#	air-script/tests/buses/buses_varlen_boundary_last.rs
#	codegen/ace/src/builder.rs
#	codegen/ace/src/tests/quotient.rs

adr1anh committed:
Address nits

adr1anh committed:
Simplify winterfell aux trace assertions codegen

adr1anh committed:
Simplify layout alignment and update comment
